src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/model/identifiers/CROIdentifier.kt

```diff
@@ -0,0 +1,91 @@
+package uk.gov.justice.digital.hmpps.personrecord.model.identifiers
+
+import org.apache.commons.lang3.builder.EqualsBuilder
+
+class CROIdentifier(inputCroId: String, inputFingerprint: Boolean, invalidInputCro: String = "") {
+
+  val croId: String = inputCroId
+  val fingerprint: Boolean = inputFingerprint
+
+  val valid: Boolean
+    get() = croId.isNotEmpty()
+
+  val invalidCro: String = invalidInputCro
+
+  override fun equals(other: Any?): Boolean {
+    return EqualsBuilder.reflectionEquals(this, other)
+  }
+
+  override fun hashCode(): Int {
+    return croId.hashCode()
+  }
+
+  override fun toString(): String {
+    return croId
+  }
+
+  companion object {
+    private const val EMPTY_CRO = ""
+    private const val SLASH = "/"
+    private const val LAST_CHARACTER = 9
+    private const val SERIAL_NUM_LENGTH = 6
+    private const val VALID_LETTERS = "ZABCDEFGHJKLMNPQRTUVWXY"
+
+    private val SF_CRO_REGEX = Regex("^SF\\d{2}/\\d{1,6}[A-Z]\$")
+    private val CRO_REGEX = Regex("^\\d{1,6}/\\d{2}[A-Z]\$")
+
+    fun from(inputCroId: String? = EMPTY_CRO): CROIdentifier {
+      if (inputCroId.isNullOrEmpty()) {
+        return CROIdentifier(EMPTY_CRO, false)
+      }
+      return toCanonicalForm(inputCroId)
+    }
+
+    private fun toCanonicalForm(inputCroId: String): CROIdentifier {
+      val (canonicalCroId, fingerprint) = when {
+        isSfFormat(inputCroId) -> Pair(canonicalizeSfFormat(inputCroId), false)
+        isExpectedFormat(inputCroId) -> Pair(canonicalizeStandardFormat(inputCroId), true)
+        else -> Pair(EMPTY_CRO, false)
+      }
+      return when {
+        (canonicalCroId.isNotEmpty() && correctModulus(canonicalCroId, fingerprint)) -> CROIdentifier(canonicalCroId, fingerprint)
+        else -> CROIdentifier(EMPTY_CRO, false, inputCroId)
+      }
+    }
+
+    private fun canonicalizeStandardFormat(inputCroId: String): String {
+      val checkChar = inputCroId.takeLast(1)
+      val (serialNum, yearPart) = inputCroId.split(SLASH) // splits into [NNNNNN, YYD]
+      val yearDigits = yearPart.dropLast(1)
+      val paddedSerialNum = padSerialNumber(serialNum)
+      return "$paddedSerialNum/$yearDigits$checkChar"
+    }
+
+    private fun canonicalizeSfFormat(inputCroId: String): String {
+      val checkChar = inputCroId.takeLast(1)
+      val (yearDigits, serialNum) = inputCroId.drop(2).split(SLASH) // splits into [YY, NNNNNND]
+      val paddedSerialNum = padSerialNumber(serialNum.dropLast(1))
+      return "$paddedSerialNum/$yearDigits$checkChar"
+    }
+
+    private fun correctModulus(inputCroId: String, fingerprint: Boolean): Boolean {
+      val checkChar = inputCroId[LAST_CHARACTER]
```

Comment:
we seem to do this three times and are often retrieving the different parts of the number. I've not thought this through, but is there anything to be said for having an interim object for validation? Something like this
```
Cro(val checkChar: String, val serialNum: String, val yearDigits: String, val fingerprint: Boolean)
```